src/wh_server_keystore.c

```diff
@@ -117,53 +165,99 @@ int hsmCacheKey(whServerContext* server, whNvmMetadata* meta, uint8_t* in)
     /* make sure id is valid */
     if (server == NULL || meta == NULL || in == NULL ||
         (meta->id & WH_KEYID_MASK) == WH_KEYID_ERASED ||
-        meta->len > WOLFHSM_CFG_SERVER_KEYCACHE_BUFSIZE) {
+        (meta->len > WOLFHSM_CFG_SERVER_KEYCACHE_BUFSIZE
+        && meta->len > WOLFHSM_CFG_SERVER_KEYCACHE_BIG_BUFSIZE
+        )) {
         return WH_ERROR_BADARGS;
     }
     /* apply client_id */
     meta->id |= (server->comm->client_id << 8);
-    for (i = 0; i < WOLFHSM_CFG_SERVER_KEYCACHE_COUNT; i++) {
-        /* check for empty slot or rewrite slot */
-        if ((foundIndex == -1 &&
-            (server->cache[i].meta->id & WH_KEYID_MASK) ==
-            WH_KEYID_ERASED) ||
-            server->cache[i].meta->id == meta->id) {
-            foundIndex = i;
+    /* check if we need to use big cache instead */
+    if (meta->len <= WOLFHSM_CFG_SERVER_KEYCACHE_BUFSIZE)
+    {
+        for (i = 0; i < WOLFHSM_CFG_SERVER_KEYCACHE_COUNT; i++) {
+            /* check for empty slot or rewrite slot */
+            if ((foundIndex == -1 &&
+                (server->cache[i].meta->id & WH_KEYID_MASK) ==
+                WH_KEYID_ERASED) ||
+                server->cache[i].meta->id == meta->id) {
+                foundIndex = i;
+            }
+        }
+        /* if no empty slots, check for a commited key we can evict */
+        if (foundIndex == -1) {
+            for (i = 0; i < WOLFHSM_CFG_SERVER_KEYCACHE_COUNT; i++) {
+                if (server->cache[i].commited == 1) {
+                    foundIndex = i;
+                    break;
+                }
+            }
+        }
+        /* write key if slot found */
+        if (foundIndex != -1) {
+            XMEMCPY((uint8_t*)server->cache[foundIndex].buffer, in, meta->len);
+            XMEMCPY((uint8_t*)server->cache[foundIndex].meta, (uint8_t*)meta,
+                sizeof(whNvmMetadata));
+            /* check if the key is already commited */
+            if (wh_Nvm_GetMetadata(server->nvm, meta->id, meta) ==
+                WH_ERROR_NOTFOUND) {
```

Comment:
Probably missing something, but how come we go to NVM here? Can we not just check `server->cache[foundIndex].commited == 1`? Seems like it would prevent a potentially lengthy search through NVM, no?

Reply:
keys are cached anytime they are read, if you look in hsmReadKey:
```
    /* cache key if free slot, will only kick out other commited keys */
    if (ret == 0 && out != NULL) {
        hsmCacheKey(server, meta, out);
    }

```
we cache a key after reading it in case it was read from NVM so we don't have to keep reading from nvm every time it's accessed since odds are if they're reading it once they'll read it again soon for more operations, thus we need to check nvm when caching to see if it's already in there

Reply:
For the nvm_flash implementation, wh_Nvm_GetMetadata copies the metadata from the directory in memory.  All of the valid metadata is read only once during init.  It shouldn't take very long to do that.